### PR TITLE
Export the input parameters used for a scenario run from the front-end

### LINKED ISSUE #148: View the input parameters used for a scenario run from the frontend
Blocked by #147 
This is the front-end task for #147

### PR DESCRIPTION
**Pull request by GRI-Admin**
DESCRIPTION OF PR:
Added PR on behalf of @Martin20494 

Closes #148 

## Developer Checklist
- [ ] Make code change
- [ ] Update tests
  - [ ] Update / create new tests
  - [ ] Ensure these tests have the expected behaviour
  - [ ] Test locally and ensure tests are passing
- [ ] Update documentation
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

## Reviewer Checklist
- [ ] Check new code for code smells
- [ ] Check new tests
  - [ ] Ensure adequate coverage
  - [ ] Check for code smells within tests
- [ ] Check if documentation needs updating
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

### CODE REVIEW

File: src/tasks.py
Comment (LukeParky) [inline src/tasks.py:1]: @Martin20494 Most of the changes in this file don't look intentional. It doesn't look like its because of rebasing, perhaps it was from copying an older version. Could you revert the changes that aren't needed, perhaps by copying the content from the most up to date ōtākaro branch. Message me if you want more info

File: src/digitaltwin/tables.py [hunk 1]
```diff
@@ -81,6 +81,42 @@ class UserLogInfo(Base):
     geometry = Column(Geometry("POLYGON", srid=2193))
 
 
+class MedusaScenarios(Base):
+    """
+    Class representing the 'MedusaUserLogInfo' table.
+
```
Comment (LukeParky) [inline src/digitaltwin/tables.py:87]: It would be good to place this new code in `pollution_tables.py` instead, because it would make it easier to pull out later

File: src/digitaltwin/tables.py [hunk 2]
```diff
@@ -81,6 +81,42 @@ class UserLogInfo(Base):
...
+    created_at : datetime
+        Timestamp indicating when the log entry was created.
+    geometry : Polygon
+        Geometric representation of the catchment area coverage.
+    """  # pylint: disable=too-few-public-methods
+
+    __tablename__ = "medusa_scenarios"
+    scenario_id = Column(Integer, primary_key=True, autoincrement=True)
+
+    antecedent_dry_days = Column(Integer, primary_key=True)
```
Comment (LukeParky) [inline src/digitaltwin/tables.py:111]: These should use `Float` instead of `Integer`

File: src/pollution_model/run_medusa_2.py [hunk 3]
```diff
@@ -625,16 +653,70 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
...
+    if scenario_id in medusa_scenarios['scenario_id'].unique():
+        return MedusaRainfallEventGeometry(
+            antecedent_dry_days=medusa_scenarios.loc[medusa_scenarios['scenario_id'] == scenario_id, 'antecedent_dry_days'].iloc[0],
+            average_rain_intensity=medusa_scenarios.loc[medusa_scenarios['scenario_id'] == scenario_id, 'average_rain_intensity'].iloc[0],
+            event_duration=medusa_scenarios.loc[medusa_scenarios['scenario_id'] == scenario_id, 'event_duration'].iloc[0],
+            rainfall_ph=medusa_scenarios.loc[medusa_scenarios['scenario_id'] == scenario_id, 'rainfall_ph'].iloc[0],
+            geometry=medusa_scenarios.loc[medusa_scenarios['scenario_id'] == scenario_id, 'geometry'].iloc[0]
+        )
+    else:
+        return None
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:707]: This is quite a complicated way because you download all of the data from the database. If you use a method similar to this snipper to BG_Flood https://github.com/GeospatialResearch/Digital-Twins/blob/7e9a0639827b8ddc19b16cbf83b3e352146f15d1/src/flood_model/bg_flood_model.py#L144-L179 where it uses a longer SQL query you can reduce the python workload.

File: src/tasks.py [hunk 4]
```diff
@@ -286,10 +297,44 @@ def get_depth_by_time_at_point(model_id: int, lat: float, lng: float) -> DepthTi
...
+            "antecedent_dry_days": float(medusa_rainfall_event.antecedent_dry_days),
+            "average_rain_intensity": float(medusa_rainfall_event.average_rain_intensity),
+            "event_duration": float(medusa_rainfall_event.event_duration),
+            "rainfall_ph": float(medusa_rainfall_event.rainfall_ph),
+            "geometry": str(medusa_rainfall_event.geometry)
+        }
+        return medusa_rainfall_dictionary
+
+    else:
+        return None
```
Comment (LukeParky) [inline src/tasks.py:330]:
```suggestion
    if medusa_rainfall_event == None:
        # The specified scenario_id does not exist
        return None
    # Write rainfall information into a dictionary
    medusa_rainfall_dictionary = {
        "antecedent_dry_days": medusa_rainfall_event.antecedent_dry_days,
        "average_rain_intensity": medusa_rainfall_event.average_rain_intensity,
        "event_duration": medusa_rainfall_event.event_duration,
        "rainfall_ph": medusa_rainfall_event.rainfall_ph,
        "geometry": medusa_rainfall_event.geometry
    }
    return medusa_rainfall_dictionary
```

Comment (LukeParky) [reply src/tasks.py:330]: These changes simplify the code flow (inverting the if) and also remove the casts that aren't necessary

File: src/pollution_model/run_medusa_2.py [hunk 5]
```diff
@@ -78,6 +80,29 @@ class MedusaRainfallEvent(NamedTuple):
...
+    rainfall_ph: float
+        The acidity level of the rainfall.
+    """
+
+    antecedent_dry_days: float
+    average_rain_intensity: float
+    event_duration: float
+    rainfall_ph: float
+    geometry: str
+
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:104]: This could be replaced with just using the `MedusaScenarios` table class, so you don't have to duplicate code

File: src/app.py [hunk 6]
```diff
@@ -284,6 +285,48 @@ def get_depth_at_point(task_id: str) -> Response:
...
+    """
+    # Get medusa rainfall information
+    medusa_rainfall_dictionary = tasks.retrieve_medusa_input_parameters.delay(scenario_id).get()
+
+    # Check if the scenario exists
+    if medusa_rainfall_dictionary is None:
+        return make_response(f"Could not find rainfall scenario {scenario_id}", NOT_FOUND)
+    else:
+        return make_response(jsonify(medusa_rainfall_dictionary), OK)
+
```
Comment (LukeParky) [inline src/app.py:326]:
```suggestion
```

File: src/app.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 from flask_swagger_ui import get_swaggerui_blueprint
 from kombu.exceptions import OperationalError
 from shapely import box
+from werkzeug.exceptions import InternalServerError
```
Comment (LukeParky) [inline src/app.py:17]: We already have `INTERNAL_SERVER_ERROR` imported
```suggestion
```

File: src/app.py [hunk 8]
```diff
@@ -284,6 +285,48 @@ def get_depth_at_point(task_id: str) -> Response:
...
+
+    Returns
+    -------
+    Response
+        Returns JSON response in the form below:
+        {
+            "antecedent_dry_days": int,
+            "average_rain_intensity": int,
+            "event_duration": int,
+            "rainfall_ph": int
```
Comment (LukeParky) [inline src/app.py:314]: This also includes geometry, which makes me think that maybe it should actually be served in geoserver. Get in contact with me to talk about this at some point this week and I can demonstrate what I mean

**Review by LukeParky — CHANGES_REQUESTED** [review]
Good additions here. I added some comments for cleanup, and make sure the linting passes. It may be more appropriate to serve the data over geoserver but that can potentially be a separate issue, let's discuss it.

File: src/pollution_model/run_medusa_2.py [hunk 9]
```diff
@@ -622,12 +626,73 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
...
+    Parameters
+    ----------
+    scenario_id: int
+        The scenario ID of the pollution model run
+
+
+    Returns
+    -------
+    LegacyRow
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:666]: This LegacyRow annotation could be replaced by a dict. Technically it's not a dict but due to python's duck typing it matches the criteria of a dict, what are your thoughts?

File: src/tasks.py [hunk 10]
```diff
@@ -286,6 +286,42 @@ def get_depth_by_time_at_point(model_id: int, lat: float, lng: float) -> DepthTi
...
+
+    Parameters
+    ----------
+    scenario_id: int
+        The scenario ID of the pollution model run
+
+
+    Returns
+    -------
+    Dict[Point, int]
```
Comment (LukeParky) [inline src/tasks.py:302]:
```suggestion
    Dict[src, Union[str, float]]
```

**Review by LukeParky — CHANGES_REQUESTED** [review]

File: src/tasks.py [hunk 11]
```diff
@@ -286,6 +286,42 @@ def get_depth_by_time_at_point(model_id: int, lat: float, lng: float) -> DepthTi
...
+            "average_rain_intensity": float(medusa_rainfall_event['average_rain_intensity']),
+            "event_duration": float(medusa_rainfall_event['event_duration']),
+            "rainfall_ph": float(medusa_rainfall_event['rainfall_ph']),
+            "geometry": str(medusa_rainfall_event['geometry'])
+        }
+
+    else:
+        medusa_rainfall_dictionary = None
+
+    return medusa_rainfall_dictionary
```
Comment (LukeParky) [inline src/tasks.py:322]:
```suggestion
    # Get rainfall scenario information
    return retrieve_input_parameters(scenario_id)
```
I think this may work, but try it out

File: src/static/api_documentation.yml [hunk 12]
```diff
@@ -310,6 +345,25 @@ components:
             type: number
           example: [ 0, 100, 200, 300, 400, 500, 600, 700, 800, 900 ]
 
+    MedusaScenarioID:
```
Comment (LukeParky) [inline src/static/api_documentation.yml:348]:
```suggestion
    MedusaScenarioInput
```

File: src/static/api_documentation.yml [hunk 13]
```diff
@@ -123,6 +126,21 @@ paths:
         '503 - Service Unavailable':
           $ref: '#/components/responses/NoCeleryWorkers'
 
+  "/scenarios/medusa/{scenario_id}":
+    get:
+      summary: Finds the depth values and corresponding time values for a particular point for a given model output task.
+      parameters:
+        - $ref: '#/components/parameters/MedusaScenarioID'
```
Comment (LukeParky) [inline src/static/api_documentation.yml:133]:
```suggestion
        - $ref: '#/components/parameters/MedusaScenarioInput'
```

File: src/static/api_documentation.yml [hunk 14]
```diff
@@ -169,6 +187,14 @@ components:
             type: string
             example: Could not find flood model output 17
 
+    MedusaScenarioIDNotFound:
+      description: The scenario with that MedusaScenarioID could not be found
+      content:
+        text/plain:
+          schema:
+            type: string
+            example: Could not find the model output 31
```
Comment (LukeParky) [inline src/static/api_documentation.yml:196]:
```suggestion
            example: Could not find rainfall scenario 31
```
I looked into this closer and it needs to match `app.py`

File: src/app.py [hunk 15]
```diff
@@ -284,6 +284,46 @@ def get_depth_at_point(task_id: str) -> Response:
...
+
+    Parameters
+    ----------
+    scenario_id: int
+        The scenario ID of the pollution model run
+
+    Returns
+    -------
+    Response
+        Returns JSON response in the form below:
```
Comment (LukeParky) [inline src/app.py:308]: Include geometry and creation times and scenarioID here

File: src/pollution_model/run_medusa_2.py [hunk 16]
```diff
@@ -622,12 +625,69 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
     rainfall_event = MedusaRainfallEvent(antecedent_dry_days, average_rain_intensity, event_duration, rainfall_ph)
 
     # Run the pollution model
-    scenario_id = run_pollution_model_rain_event(engine, area_of_interest, rainfall_event)
+    scenario_id = run_pollution_model_rain_event(
+        engine, area_of_interest, rainfall_event
+    )  # pylint: disable=unsupported-assignment-operation
+
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:631]: I don't think these changes are needed, my pylint didn't fail here
```suggestion
    scenario_id = run_pollution_model_rain_event(engine, area_of_interest, rainfall_event)
```

File: src/pollution_model/run_medusa_2.py [hunk 17]
```diff
@@ -622,12 +625,69 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
...
+    query = MedusaScenarios(
+        scenario_id=scenario_id,
+        antecedent_dry_days=antecedent_dry_days,
+        average_rain_intensity=average_rain_intensity,
+        event_duration=event_duration,
+        rainfall_ph=rainfall_ph,
+        geometry=area_of_interest['geometry'].to_wkt().iloc[0]
+    )
+    # Execute the query
+    execute_query(engine, query)
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:647]: query is very vague here
```suggestion
    # Create new table recording users' history
    create_table(engine, MedusaScenarios)
    # Record the input parameters used to create the scenario
    record_scenario_input_query = MedusaScenarios(
        scenario_id=scenario_id,
        antecedent_dry_days=antecedent_dry_days,
        average_rain_intensity=average_rain_intensity,
        event_duration=event_duration,
        rainfall_ph=rainfall_ph,
        geometry=area_of_interest['geometry'].to_wkt().iloc[0]
    )
    # Execute the query
    execute_query(engine, record_scenario_input_query)
```

File: src/pollution_model/run_medusa_2.py [hunk 18]
```diff
@@ -622,12 +625,69 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
...
+
     return scenario_id
 
 
+def retrieve_input_parameters(scenario_id: int) -> Dict[str, Union[str, float]]:
+    # pylint: disable=unsupported-assignment-operation
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:653]:
```suggestion
```
It's not recommended to disable this for the entire function, its a useful warning that we should not fully ignore.
Let me know if this causes annoyance in programming though, because we can reconsider

File: src/pollution_model/run_medusa_2.py [hunk 19]
```diff
@@ -622,12 +625,69 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
...
+    Parameters
+    ----------
+    scenario_id: int
+        The scenario ID of the pollution model run
+
+
+    Returns
+    -------
+    Dict[str, Union[str, float]]
+        A dictionary with information selected from Rainfall MEDUSA 2.0 database based on scenario ID
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:666]: After we found out that LegacyRow did not serialise correctly in task.py I've found out that its not right to have this return type.

Two options for fixing:
1. have this return a `dict` using `dict(row)`

2. change the annotation back to `LegacyRow`

File: src/run_all.py [hunk 20]
```diff
@@ -50,54 +44,54 @@ def main(
...
+    #     "interval_mins": 10,
+    #     "proj_year": 2030,
+    #     "confidence_level": "low",
+    #     "ssp_scenario": "SSP1-2.6",
+    #     "add_vlm": False,
+    #     "percentile": 50,
+    #     "log_level": LogLevel.INFO
+    # },
+    # main_river: {
+    #     "flow_length_mins": 2880,
```
Comment (LukeParky) [inline src/run_all.py:73]: restore all of these

File: src/tasks.py [hunk 21]
```diff
@@ -286,6 +286,32 @@ def get_depth_by_time_at_point(model_id: int, lat: float, lng: float) -> DepthTi
     return DepthTimePlot(depths, times)
 
 
+@app.task(base=OnFailureStateTask)
+def retrieve_medusa_input_parameters(scenario_id: int) -> Dict[str, Union[str, float]]:
```
Comment (LukeParky) [inline src/tasks.py:290]: this could also return `None`, so it needs to be wrapped in an `Optional[ ]` so that its clear. It would be good to describe in the docstring what `None` means

File: src/tasks.py [hunk 22]
```diff
@@ -286,6 +286,32 @@ def get_depth_by_time_at_point(model_id: int, lat: float, lng: float) -> DepthTi
...
+    """
+    Retrieve input parameters for the current scenario id. This is used for app file.
```
Comment (LukeParky) [inline src/tasks.py:292]:
```suggestion
    Retrieve input parameters for the current scenario id.
```

**Review by LukeParky — CHANGES_REQUESTED** [review]
Looking overall good, but I did notice a few things that I'd like changed so we don't have to come back to it in the future

File: src/run_all.py [hunk 23]
```diff
@@ -97,7 +97,7 @@ def main(
         "antecedent_dry_days": 1,
         "average_rain_intensity": 1,
         "event_duration": 1,
-        "rainfall_ph": 7
+        "rainfall_ph": 8
```
Comment (LukeParky) [inline src/run_all.py:100]:
```suggestion
        "rainfall_ph": 7
```
Keeping the PR changes as small as possible is good, so I am suggesting changing this back to 7

File: src/pollution_model/run_medusa_2.py [hunk 24]
```diff
@@ -622,12 +625,66 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame,
     rainfall_event = MedusaRainfallEvent(antecedent_dry_days, average_rain_intensity, event_duration, rainfall_ph)
 
     # Run the pollution model
-    scenario_id = run_pollution_model_rain_event(engine, area_of_interest, rainfall_event)
+    scenario_id = scenario_id = run_pollution_model_rain_event(engine, area_of_interest, rainfall_event)
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:628]:
```suggestion
    scenario_id = run_pollution_model_rain_event(engine, area_of_interest, rainfall_event)
```

File: src/tasks.py [hunk 25]
```diff
@@ -286,6 +286,32 @@ def get_depth_by_time_at_point(model_id: int, lat: float, lng: float) -> DepthTi
...
+    Dict[src, Union[str, float]]
+        A dictionary contain information from Rainfall MEDUSA 2.0 database or None if scenario does not exist.
+    """
+    # Get rainfall information
+    medusa_rainfall_event = retrieve_input_parameters(scenario_id)
+
+    # Return a dictionary format of these parameters or None if no ID found
+    if medusa_rainfall_event is None:
+        return None
+
```
Comment (LukeParky) [inline src/tasks.py:311]: I couldn't get this part of the code to run. If I put in an invalid scenario_id like `9999` into `retrieve_input_parameters` then it raises a `TypeError`. Your documentation says it should raise `FileNotFoundError` and this section of code suggests it should return `None`.

I like the idea of it returning `None` but the implementation and documentation of `retrieve_input_parameters` isn't consistent yet

**Review by LukeParky — CHANGES_REQUESTED** [review]

File: src/pollution_model/run_medusa_2.py [hunk 26]
```diff
@@ -661,28 +661,28 @@ def retrieve_input_parameters(scenario_id: int) -> Dict[str, Union[str, float]]:
...
+
+        # Get information by using scenario_id from medusa_scenarios table in the dataset
+        row = engine.execute(query).fetchone()
 
-    return row
+        # Return
+        if row is None:
+            return None
+        else:
+            return dict(row)
```
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:685]: Remove all the `else` here. It simplifies he level of indentation and makes the code easier to visually read.
The return is instant so the `else` will never be hit if it returns

Comment (Martin20494) [reply src/pollution_model/run_medusa_2.py]: I have removed them.

File: src/tasks.py (hunk 21 above)
Comment (Martin20494) [reply src/tasks.py]: I have added `Optional[ ]` in.

**Review by LukeParky — APPROVED** [review]
